### PR TITLE
Task7

### PR DESCRIPTION
**Pull request by rayblair06**
Task 7.1
- Install and configure `Jest`
- Separated the functionality for listening to `server.js` for easier testing
- Added Integration Tests for `Group` and `User` routes and Unit Tests for `Group` and `User` services.
- Mocked model functionality. (I was unable to successfully mock sequelize.)  

Task 7.2
Added an `.env` and `dotenv` in a previous task so I believe this task is already complete.

Additional
- Default database env variables
- Added `.dockerignore`
- Added `node_modules` volume to docker-compose.
- Added a 'base' route so app returns 200 on root.

### CODE REVIEW

File: tests/feature/controllers/group.test.js [hunk 1]
```diff
@@ -0,0 +1,265 @@
+import Group from '../../../src/db/models/group';
+import { GroupNotFoundError } from '../../../src/exceptions/errors';
+import * as AuthService from '../../../src/services/auth';
+import * as GroupService from '../../../src/services/group';
+
+
+const supertest = require('supertest');
```
Comment (Dmitry-White) [inline tests/feature/controllers/group.test.js:7]: Consider using the same module system for all imports in your project.

File: tests/feature/controllers/user.test.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
+import User from '../../../src/db/models/user';
+import { UserNotFoundError } from '../../../src/exceptions/errors';
+import * as AuthService from '../../../src/services/auth';
+
+
+const supertest = require('supertest');
+
+const { app } = require('../../../src/app');
```
Comment (Dmitry-White) [inline tests/feature/controllers/user.test.js:8]: Same as above.

File: tests/feature/controllers/group.test.js [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    age: 21,
+    isDeleted: false
+});
+
+describe('Test the groups controller', () => {
+    test('It should return all groups', async () => {
+        // Mock Model findAll
+        jest.spyOn(Group, 'findAll').mockImplementation(() => {
+            return Promise.resolve([
+                {
```
Comment (Dmitry-White) [inline tests/feature/controllers/group.test.js:25]: I can see that these objects are the same across several tests.
Consider always preparing test dummies before at the beginning of the file and reusing them.

File: tests/feature/controllers/base.test.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+const { app } = require('../../../src/app');
+
+
+describe('Test the root path', () => {
+    test('It should response the GET method', async () => {
+        const response = await supertest(app)
+            .get('/');
+
+        expect(response.statusCode).toBe(200);
```
Comment (Dmitry-White) [inline tests/feature/controllers/base.test.js:11]: Consider using the "http-status-code" library that you've added to the business logic already, that was the main purpose actually :)

File: tests/feature/controllers/user.test.js [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    login: 'John Doe',
+    password: 'secret',
+    age: 21,
+    isDeleted: false
+});
+
+describe('Test the users controller', () => {
+    test('It should return all users', async () => {
+        // Mock Model findAll
+        jest.spyOn(User, 'findAll').mockImplementation(() => {
```
Comment (Dmitry-White) [inline tests/feature/controllers/user.test.js:22]: Consider extracting all the `jest.spyOn`s mocking logic into a utils file a the `tests/` level.
You can write some wrapper function around `jest.spyOn` that will expect a method call and a returned value as arguments. That way each call to that wrapper function will have its on stuff mocked in closure.

Comment (Dmitry-White) [reply tests/feature/controllers/user.test.js:22]: And drastically reduce repetition of mocking logic :)

File: tests/unit/services/group.test.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+                    'id': '79836ba5-e47f-4322-a498-53e7744189ad',
+                    'name': 'Read only',
+                    'permissions': [
+                        'READ'
+                    ]
+                }
+            );
+        });
+
+        const output = await GroupService.findById('79836ba5-e47f-4322-a498-53e7744189ad');
```
Comment (Dmitry-White) [inline tests/unit/services/group.test.js:57]: Consider using constants and functions to reduce repeated values and object and reuse as much as possible.

File: tests/unit/services/user.test.js [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                'success': false,
+                'status_code': 404,
+                'errors': {
+                    'message': 'User Not Found'
+                }
+            })
+        );
+    });
+
+    test('It can suggest usernames', async () => {
```
Comment (Dmitry-White) [inline tests/unit/services/user.test.js:103]: Also, try using `test.each` and `describe.each` for tests with repeated structure

**Review by Dmitry-White — CHANGES_REQUESTED** [review]

**Comment by rayblair06** [comment]
Thanks for taking the time to review my PR @Dmitry-White

I have made the following changes based on your advice.
- Replaced all `require` with `import`.
- `supertest` is now within its own `request` mock module.
- Wrapper module for `jest.spyOn`.
- Model 'Fakes' will mock the called function, if data is passed it will use that else it will use 'Dummie' data.
- TestSuite is now initialised with model Fake.
- Used `test.each` to provide more testcases for 'It can suggest username' test.
- General refactoring and clean up.

**Review by Dmitry-White — APPROVED** [review]
Looks awesome!

**Pull request merged**